_posts/2021-01-04-Designing-REST-services.md

```diff
@@ -0,0 +1,331 @@
+* Last-Modified: The date at which the resource was modified the last time before this request.
+
+### Consistency
+
+It is important that data is kept consistent across various manipulations, especially when it is being exposed to and manipulated by the consumers of stateless services. 
+One of the key risks that threaten this consistency is concurrent writes to the same resource. 
+The scenario is straightforward. Consumer A reads the current resource. Consumer B reads the same resource. 
+Consumer B updates the resource, followed by consumer A updating the resource. 
+In this scenario, consumer A never saw the changes that consumer B made, and more than likely erases them by updating the resource.
+
+One way to tackle this scenario is to introduce conditional updates. 
+The idea is that when a consumer reads a certain resource, it gets passed a unique representation of this resource in the “Etag”-header (for example a MD5-hash of the non-transient fields of the resource). 
+Each time any consumer is updating a resource, the consumer needs to provide this value in the “If-Match”-header. 
+The idea is that when the update is processed by the service, it checks whether the content of the header still matches the representation of the resource. 
+If it is still the same, it will update the resource and return the result (HTTP 200 OK status). 
+If between the read and update of the resource by this consumer, another consumer modified the resource, the representation value will have changed, and the update is denied (HTTP 412 PRECONDITION FAILED status). 
+It is then up to the consumer to fetch an up to date resource and decide whether or not still to process its changes. 
+The “If-None-Match”-header checks whether a resource exists with the representation, and if not, proceed with the update. 
+This could be used to prevent the creation of duplicate resources in the database.
+
+{:refdef: style="text-align: center;"}
+<img src="{{ '/img/2021-01-04-Designing-REST-services/etag.png' | prepend: site.baseurl }}" alt="Workspace" class="image" style="margin:0px auto; max-width:100%">
+{: refdef}
+
+There are also headers based on the modification date of a resource for indicating to the service that it should behave in a conditional way. 
+Personally, I feel that this adds the additional complication of syncing the date between the service and its consumers, and that using the ETag is more resilient:
+* The “If-Modified-Since” header only executes when the resource has been modified since the given date. This header works with the last-modified header instead of the ETag.
+* The “If-Unmodified-Since” header is the inverse of the “If-Modified-Since” header.
+
+### Security
+
+There is an entire chapter dedicated to security in the book. 
+However, since this is one of those areas of expertise that has rapid evolutions the contents of this book (written in 2015) are somewhat dated. 
```

Comment:
```suggestion
However, since this is one of those areas of expertise that has rapid evolutions, the contents of this book (written in 2015) are somewhat dated. 
```

Reply:
done